var/spack/repos/builtin/packages/opencv/package.py

```diff
@@ -2,399 +2,1062 @@
+        conflicts("~photo")
 
     # Optional 3rd party components (variants)
     # Defined in `CMakeLists.txt` and `modules/gapi/cmake/init.cmake`
     # using `OCV_OPTION(WITH_* ...)`
     components = [
-        '1394', 'ade', 'android_mediandk', 'android_native_camera', 'aravis',
-        'avfoundation', 'cap_ios', 'carotene', 'clp', 'cpufeatures', 'cublas', 'cuda',
-        'cudnn', 'cufft', 'directx', 'dshow', 'eigen', 'ffmpeg', 'freetype', 'gdal',
-        'gdcm', 'gphoto2', 'gstreamer', 'gtk', 'gtk_2_x', 'halide', 'hpx',
-        'imgcodec_hdr', 'imgcode_pfm', 'imgcodec_pxm', 'imgcodec_sunraster',
-        'inf_engine', 'ipp', 'itt', 'jasper', 'jpeg', 'lapack', 'librealsense', 'mfx',
-        'msmf', 'msmf_dxva', 'ngraph', 'nvcuvid', 'onnx', 'opencl', 'openclamdblas',
-        'openclamdfft', 'opencl_d3d11_nv', 'opencl_svm', 'openexr', 'opengl',
-        'openjpeg', 'openmp', 'openni', 'openni2', 'openvx', 'plaidml', 'png',
-        'protobuf', 'pthreads_pf', 'pvapi', 'qt', 'quirc', 'tbb', 'tengine', 'tiff',
-        'ueye', 'v4l', 'va', 'va_intel', 'vtk', 'vulcan', 'webp', 'win32ui', 'ximea',
-        'xine'
+        "1394",
+        "ade",
+        "android_mediandk",
+        "android_native_camera",
+        "avfoundation",
+        "cap_ios",
+        "carotene",
+        "clp",
+        "cpufeatures",
+        "cublas",
+        "cuda",
+        "cudnn",
+        "cufft",
+        "directx",
+        "dshow",
+        "eigen",
+        "ffmpeg",
+        "gdal",
+        "gtk",
+        "hpx",
+        "imgcodec_hdr",
+        "imgcodec_pfm",
+        "imgcodec_pxm",
+        "imgcodec_sunraster",
+        "ipp",
+        "itt",
+        "jasper",
+        "jpeg",
+        "lapack",
+        "msmf",
+        "msmf_dxva",
+        "onnx",
+        "opencl",
+        "opencl_d3d11_nv",
+        "openexr",
+        "opengl",
+        "openjpeg",
+        "openmp",
+        "openvx",
+        "plaidml",
+        "png",
+        "protobuf",
+        "pthreads_pf",
+        "qt",
+        "quirc",
+        "tbb",
+        "tengine",
+        "tesseract",
+        "tiff",
+        "v4l",
+        "vtk",
+        "vulcan",
+        "webp",
+        "win32ui",
     ]
 
+    # These likely need additional spack packages
+    components_pending = [
+        "aravis",
+        "gdcm",
+        "gphoto2",
+        "gstreamer",
+        "gtk_2_x",  # deprecated in spack
+        "halide",
+        "inf_engine",
+        "librealsense",
+        "mfx",
+        "ngraph",
+        "nvcuvid",  # disabled, details: https://github.com/opencv/opencv/issues/14850
+        "opencl_svm",
+        "openclamdblas",
+        "openclamdfft",
+        "openni",
+        "openni2",
+        "pvapi",
+        "ueye",
+        "va",
+        "va_intel",
+        "ximea",
+        "xine",
+    ]
+
+    # components and modules with the same name
+    # used in `def cmake_args(self)`
+    component_and_module = ["freetype", "julia", "matlab"]
```

Comment:
Can you clarify this? For these components/modules, is it possible to enable one and disable the other in CMake? If so, we should find a way to do this in the Spack package. If we have to, we can convert components/modules/etc to multi-valued variants to avoid naming conflict.

Reply:
I structured this such that the variant is for the opencv module, and the supporting component, or external library, is set based on that. For instance, if the freetype variant is selected that does the following:
- sets the freetype opencv module to be built
- sets a dependency on freetype
- ensures that the cmake flag to use the spack freetype package is set

It seems reasonable to tie those together, as we want to use the spack freetype package if the opencv freetype module is built. Likewise, there would be no point in including a spack freetype package if the opencv freetype module is not built. I do not see how multi-valued variants would solve this, and in fact, I believe would make it more confusing for someone installing opencv and looking at the wall of variants. My goal was to present only one freetype variant, not two.